### ISSUE TITLE
Value Checker issues warnings when casting between integral types with range annotations

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#1264]
Consider the following code:

```java
byte foo(@IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE) int x) {
    return (byte) x;
}
```

The Value Checker issues a warning that the cast from `@IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE)` to `@IntRange(-128, 127)` is unsafe. This warning is a false positive.

The same problem occurs whenever there is a downcast between integral types, not just between `int` and `byte`

**Comment by smillst** [comment]
When this issue is fixed, this daikon change should be reverted:
https://github.com/codespecs/daikon/commit/2ac5315e02aa40d042686011d3ed53bf9b8ba86a

**Comment by smillst** [comment]
Implementation note: 

The ValueVisitor should probably override org.checkerframework.common.basetype.BaseTypeVisitor#checkTypecastSafety and checkTypecastRedundancy.

**Comment by wmdietl** [comment]
Out of curiosity: why is the cast a false positive? The cast goes from a larger range to a smaller range, which would seem like exactly what the Value Checker should warn about.

**Comment by smillst** [comment]
The cast is from an int to a byte.  A byte's range is at most -128 to 127. So casting from `@IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE) int` to `@IntRange(-128, 127) byte` is always safe.

**Comment by wmdietl** [comment]
I would agree that casting from `@IntRange(-128, 127) int` to `@IntRange(-128, 127) byte` is safe.
But casting from all possible integers to `@IntRange(-128, 127) byte` looses information. I must be missing something.

**Comment by smillst** [comment]
What Value Checker type should '(byte) x' have if not '@IntRange(-128, 127)'?

**Comment by kelloggm** [comment]
Or, put another way, even if the integer is out of the range of the byte type, the resulting byte will still have type `@IntRange(-128, 127)` because the value will overflow, and a byte cannot hold values outside that range.

**Comment by wmdietl** [comment]
Do you get a warning when you assign/cast a `@IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE) int` to a `@IntRange(-128, 127) int`?

**Comment by smillst** [comment]
Yes, `(@IntRange(-128, 127) int) x`should issue a warning, assuming the type of `x` is `@IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE) int`.  Because the value of `(@IntRange(-128, 127) int) x` could still be outside that range.

**Comment by wmdietl** [comment]
To me it would seem useful if the Value Checker gives a warning when a larger range is assigned to a smaller one. This is one of the tricky situations when handling primitives and I would be surprised to not get a warning here.
Your argument is that the underlying byte isn't storing a value outside of -128 to 127 and it's Java that is truncating the value.
I would still rather expect a warning from the Value Checker that this is happening than silently ignoring it.

**Comment by smillst** [comment]
At narrowing primitive conversions, the Value Checker could issue a warning if the value being converted is not within the range of the new primitive type.  The problem is that in practice such a warning is usually a false positive, because if the user wrote such a cast, they expected the lossy behavior.

**Comment by wmdietl** [comment]
I'm fine with the argument that the developer already had to write an explicit cast. I tripped over the "The Value Checker issues a warning that the cast from @IntRange(from = Integer.MIN_VALUE, to = Integer.MAX_VALUE) to @IntRange(-128, 127) is unsafe." description in the original report, where just the ranges are listed.
A note about what the checker guarantees might still be a nice addition to the manual.

**Comment by mernst** [comment]
Here is a proposed resolution, which I think we all agree with:
 * don't issue a warning at narrowing numeric casts.  Assume that if the programmer explicitly wrote such a cast, it was intentional and the programmer meant to lose precision.
 * In the future, consider adding a mode, controlled by a command-line option, in which the programmer is warned about such loss of precision.  There is no need to implement this yet; it could be a TODO comment in the code.
